src/shutdown.rs

```diff
@@ -1,66 +1,116 @@
 }
 
 impl Shutdown {
-    pub fn new() -> Self {
-        let (sender, receiver) = watch::channel(false);
-        Self { sender, receiver }
+    /// Create a new shutdown and spawn a task to monitor it.
+    ///
+    /// * `timeout` - The maximum time to wait for all receivers to be dropped.
+    ///               If all receivers have not been dropped by the timout the
+    ///               process will exit with code 1.
+    ///
+    /// * `delay` - A minimum delay before calling [`std::process::exit`]. This us useful to allow
+    ///             for cancaled futures to make it to an await point.
+    ///
+    /// Shutdown can be triggered by calling [`Shutdown::shutdown`].
+    /// The shutdown will begin immediately after the delay expires if no receivers
+    /// are being held.
+    pub fn spawn(timeout: Duration, delay: Duration) -> Self {
+        let (sender, _) = watch::channel(false);
+        let shutdown = Self { sender };
+        shutdown.clone().spawn_monitor(timeout, delay);
+        shutdown
     }
 
     /// Send the signal to shutdown the program.
     pub fn shutdown(&self) {
-        // Does not fail because the channel never closes.
-        self.sender.send(true).unwrap();
+        self.sender.send(true).ok();
     }
 
     /// Are we currently shutting down?
     #[must_use]
     pub fn is_shutting_down(&self) -> bool {
-        *self.receiver.borrow()
+        *self.sender.subscribe().borrow()
+    }
+
+    /// Returns a `Receiver` which must be held until the caller is ready to shutdown.
+    #[must_use]
+    pub fn handle(&self) -> Receiver<bool> {
+        self.sender.subscribe()
     }
 
-    /// Wait for the program to shutdown.
+    /// Wait for the shutdown to begin.
     ///
-    /// Resolves immediately if the program is already shutting down.
-    /// The resulting future is safe to cancel by dropping.
-    pub fn await_shutdown(&self) -> impl Future<Output = ()> + 'static {
-        let mut watch = self.receiver.clone();
-
-        async move {
-            if *watch.borrow_and_update() {
-                return;
-            }
-            // Does not fail because the channel never closes test_config.
-            watch.changed().await.unwrap();
+    /// Returns a `Receiver` which must be held until the caller is ready to shutdown.
+    #[must_use]
+    pub async fn await_shutdown_begin_with_handle(&self) -> Receiver<bool> {
+        let mut receiver = self.sender.subscribe();
+        if *receiver.borrow_and_update() {
+            return receiver;
         }
+        receiver.changed().await.ok();
+        receiver
     }
-}
 
-impl Default for Shutdown {
-    fn default() -> Self {
-        Self::new()
+    /// Wait for the shutdown to begin.
+    ///
+    /// If the caller down not already own a `Receiver`
+    /// then the process may abort immediately when this returns.
+    pub async fn await_shutdown_begin(&self) {
+        let _ = self.await_shutdown_begin_with_handle().await;
```

Comment:
```suggestion
        let _ = self.await_shutdown_begin_with_handle().await.await;
```
?

As it stands the type of `_` is `Receiver<bool>`

Reply:
I'm not sure what you're suggesting I don't think? This looks as intended to me. `Receiver<boo>` is `must_use` so I have to access it like this to throw it away. You also cannot `await` a `Receiver`.